spec/03-plugins/25-oauth2/03-access_spec.lua

```diff
@@ -2905,6 +2950,97 @@ describe("Plugin: oauth2 [#" .. strategy .. "]", function()
       end)
+      describe("refreshing token", function()
+        local request_client, token
+        it("fails when scope is mismatching", function ()
+          -- provision code
+          local code, body, res
+          request_client = helpers.proxy_ssl_client()
+          body = {
+              provision_key = "provision123",
+              client_id = "clientid123",
+              response_type = "code",
+              scope = "scope18",
+              state = "hello",
+              authenticated_userid = "userid123",
+          }
+          res = assert(request_client:send {
+            method = "POST",
+            path = "/oauth2/authorize",
+            body = body,
+            headers = kong.table.merge({
+              ["Host"] = "oauth2_18.com",
+              ["Content-Type"] = "application/json"
+            })
+          })
+          res = assert(cjson.decode(assert.res_status(200, res)))
+          if res.redirect_uri then
+            local iterator, err = ngx.re.gmatch(res.redirect_uri, "^http://google\\.com/kong\\?code=([\\w]{32,32})&state=hello$")
+            assert.is_nil(err)
+            local m, err = iterator()
+            assert.is_nil(err)
+            code = m[1]
+          end
+
+          -- provision token
+          body = {
+            code = code,
+            client_id = "clientid123",
+            client_secret = "secret123",
+            grant_type = "authorization_code",
+            redirect_uri = "http://google.com/kong",
+          }
+          res = assert(request_client:send {
+            method  = "POST",
+            path    = "/oauth2/token",
+            body    = body,
+            headers = {
+              ["Host"]         = "oauth2_18.com",
+              ["Content-Type"] = "application/json"
+            }
+          })
+          token = assert(cjson.decode(assert.res_status(200, res)))
+
+          -- refresh token with mismatching scope
+          res = assert(request_client:send {
+            method  = "POST",
+            path    = "/oauth2/token",
+            body    = {
+              refresh_token    = token.refresh_token,
+              client_id        = "clientid123",
+              client_secret    = "secret123",
+              grant_type       = "refresh_token",
+            },
+            headers = {
+              ["Host"]         = "oauth2_19.com",
+              ["Content-Type"] = "application/json"
+            }
+          })
+          res = assert(cjson.decode(assert.res_status(400, res)))
+          assert.same({
+            error = "invalid_scope",
+            error_description = "scope mismatch"
+          }, res)
+        end)
+        it("successes when scope is a subset", function()
+          -- refresh token with mismatching scope
+          local res = assert(request_client:send {
+            method  = "POST",
+            path    = "/oauth2/token",
+            body    = {
+              refresh_token    = token.refresh_token,
```

Comment:
It seems that this test reuses a token generated by a previous test.  Please make each test self contained.  The outcome of one test should not influence other tests, as that makes it hard to run and debug tests in isolation.  Also, it is not clear from looking at the test how the token has been generated, making it difficult to understand what is being tested.

Reply:
addressed